### ISSUE TITLE
Monitoring - Prometheus with persistent storage fails to start with CrashLoopBackOff

### ISSUE THREAD

**Issue by zidex** [rancher/rancher#17256]
**What kind of request is this (question/bug/enhancement/feature request):**
bug

**Environment information**
rancher/rancher:v2.2.0-alpha3
k8s v1.12.3

**Steps to reproduce (least amount of steps as possible):**
Enable cluster monitoring
Enable Persistent Storage for Prometheus

**Result:**
Prometheus container fails to start with:

> CrashLoopBackOff: Back-off 5m0s restarting failed container=prometheus  pod=prometheus-cluster-monitoring-0_cattle-prometheus(1f55df92-0a83-11e9-a30c-902b34d02d5c)

Prometheus container log:

> level=error ts=2018-12-28T10:52:14.870125298Z caller=main.go:625 err="opening storage failed: create dir: mkdir /prometheus/wal: permission denied"

Additional info:

Prometheus without persistent storage starts successfully
Grafana persistent storage is functional

**Comment by zidex** [comment]
I use local storage with Local Node Path plugin

**Comment by thxCode** [comment]
@zidex , I cannot reproduce that. Can you descript more about which environment you were deploying your cluster? Is it a Cloud or Cloud Kubernetes Service or something else?

**Comment by thxCode** [comment]
@zidex , FYI

Starting Prometheus Operator v0.23.2 with `--disable-auto-user-group=false` (using in Rancher monitoring preview version) will inject the 'correct' `SecurityContext`:

https://github.com/coreos/prometheus-operator/blob/27b1eb72d9d93e5ab447121e2b884bc558bde01d/pkg/prometheus/statefulset.go#L334-L340

Entering `prometheus-cluster-monitoring` Pod:
```
/prometheus $ id
uid=1000 gid=0(root) groups=2000
/prometheus $ ls -al wal
total 24188
drwxr-xr-x    2 1000     root          4096 Jan  4 08:29 .
drwxrwxrwx    3 root     root          4096 Jan  4 08:29 ..
-rw-r--r--    1 1000     root      24758420 Jan  4 08:37 00000000
/prometheus $ ls -al
total 12
drwxrwxrwx    3 root     root          4096 Jan  4 08:29 .
drwxr-xr-x    1 root     root          4096 Jan  4 08:29 ..
drwxr-xr-x    2 1000     root          4096 Jan  4 08:29 wal
```

However in v0.26.0, it becomes user customization:

https://github.com/coreos/prometheus-operator/blob/77764ee4239f2a8a7917883651f9e92ea3c52924/pkg/prometheus/statefulset.go#L367-L370

Entering `prometheus-cluster-monitoring` Pod:
```
/prometheus $ id
uid=65534(nobody) gid=65534(nogroup)
/prometheus $ ls
wal
/prometheus $ ls -al
total 12
drwxrwxrwx    3 root     root          4096 Jan  4 09:26 .
drwxr-xr-x    1 root     root          4096 Jan  4 09:26 ..
drwxr-xr-x    2 nobody   nogroup       4096 Jan  4 09:26 wal
/prometheus $ ls -al wal
total 15844
drwxr-xr-x    2 nobody   nogroup       4096 Jan  4 09:26 .
drwxrwxrwx    3 root     root          4096 Jan  4 09:26 ..
-rw-r--r--    1 nobody   nogroup   16212770 Jan  4 09:28 00000000
```

So if you using preview version, don't upgrade Prometheus Operator version. The workaround is dropping the stale Prometheus PVC.

**Comment by thxCode** [comment]
Oh, something mistake by me, I found a similar issue: https://github.com/coreos/prometheus-operator/issues/830#issuecomment-379163035.

**Comment by thxCode** [comment]
@zidex , I think `kubernetes.io/no-provisioner` Storage Provisioner isn't respecting the `SecurityContext`, you should grant the appropriate permissions manually. On your steps, you can `chown -R 1000:2000 /mnt/drivers/prometheus` in the `test1604` host. 

We will provide a configurable `SecurityContext` after https://github.com/rancher/system-charts/pull/8 merged.

**Comment by zidex** [comment]
The directory **/mnt/disks/prometheus** is not created when monitoring is enabled. I created it with my hands and made **chown -R 1000:2000**. The problem did not disappear.

I looked at the list of volumes for Pod prometheus-cluster-monitoring, the required volume is missing there. There is only: config, config-out, prometheus-cluster-monitoring-rulefiles-0, secret-exporter-etcd-cert

If you enable data storage for Grafana, the directory is created, but with incorrect permissions. After chmod 777, everything starts working.

**Comment by thxCode** [comment]
@zidex , `kubernetes.io/no-provisioner` should never create a subpath dir: `/prometheus` for you, this volume mechanism is managed by Prometheus Operator, you can check the `spec` of deployed Prometheus StatefulSet.

Back to the actual problem, as you can see on https://github.com/coreos/prometheus-operator/issues/830#issuecomment-379163035, it causes by the Storage Provisioner which doesn't or can't respect the `SecurityContext`. You can check the application image Dockerfile to ensure running permissions.

Grafana is setting as below, [reference](http://docs.grafana.org/installation/docker/#reading-secrets-from-files-support-for-docker-secrets):
```yaml
securityContext:
    fsGroup: 472
``` 

Prometheus and Alertmanager will set as below:
```yaml
securityContext:
    runAsUser: 1000
    runAsNonRoot: true
    fsGroup: 2000
```

**Comment by thxCode** [comment]
@zidex, is there any other problem with this?

**Comment by loganhz** [comment]
Please let us know in a comment if you can still reproduce the issue, and we'll reopen it

**Comment by danielhass** [comment]
I'm still experiencing this issue. After activating the Cluster Monitoring the following errors pop up in the prometheus container log:
```
level=warn ts=2019-07-05T09:30:32.887785616Z caller=main.go:295 deprecation_notice="\"storage.tsdb.retention\" flag is deprecated use \"storage.tsdb.retention.time\" instead."
level=info ts=2019-07-05T09:30:32.887845114Z caller=main.go:302 msg="Starting Prometheus" version="(version=2.7.1, branch=HEAD, revision=62e591f928ddf6b3468308b7ac1de1c63aa7fcf3)"
level=info ts=2019-07-05T09:30:32.887865005Z caller=main.go:303 build_context="(go=go1.11.5, user=root@f9f82868fc43, date=20190131-11:16:59)"
level=info ts=2019-07-05T09:30:32.887883431Z caller=main.go:304 host_details="(Linux 4.4.162-94.72-default #1 SMP Mon Nov 12 18:57:45 UTC 2018 (9de753f) x86_64 prometheus-cluster-monitoring-0 (none))"
level=info ts=2019-07-05T09:30:32.887901595Z caller=main.go:305 fd_limits="(soft=1048576, hard=1048576)"
level=info ts=2019-07-05T09:30:32.887917069Z caller=main.go:306 vm_limits="(soft=unlimited, hard=unlimited)"
level=info ts=2019-07-05T09:30:32.894063049Z caller=main.go:620 msg="Starting TSDB ..."
level=info ts=2019-07-05T09:30:32.894324467Z caller=main.go:489 msg="Stopping scrape discovery manager..."
level=info ts=2019-07-05T09:30:32.894338028Z caller=main.go:503 msg="Stopping notify discovery manager..."
level=info ts=2019-07-05T09:30:32.894343901Z caller=main.go:525 msg="Stopping scrape manager..."
level=info ts=2019-07-05T09:30:32.894351197Z caller=main.go:499 msg="Notify discovery manager stopped"
level=info ts=2019-07-05T09:30:32.894369838Z caller=main.go:485 msg="Scrape discovery manager stopped"
level=info ts=2019-07-05T09:30:32.894382328Z caller=manager.go:736 component="rule manager" msg="Stopping rule manager..."
level=info ts=2019-07-05T09:30:32.894394025Z caller=manager.go:742 component="rule manager" msg="Rule manager stopped"
level=info ts=2019-07-05T09:30:32.894402025Z caller=notifier.go:521 component=notifier msg="Stopping notification manager..."
level=info ts=2019-07-05T09:30:32.894409066Z caller=main.go:679 msg="Notifier manager stopped"
level=info ts=2019-07-05T09:30:32.894422316Z caller=main.go:519 msg="Scrape manager stopped"
level=info ts=2019-07-05T09:30:32.894454406Z caller=web.go:416 component=web msg="Start listening for connections" address=127.0.0.1:9090
level=error ts=2019-07-05T09:30:32.894921303Z caller=main.go:688 err="opening storage failed: create dir: mkdir /prometheus/wal: permission denied"
```

Running on Rancher v2.2.4 with Custom Nodes.
Prom Container image: rancher/prom-prometheus:v2.7.1
Persistent storage for Monitoring: error happens with disabled and enabled.

Any ideas on this?

**Comment by loganhz** [comment]
@thxCode ^^

**Comment by adeleglise** [comment]
Hi, I'm facing the same issue, using latest stable rancher and monitoring chart 0.0.3.

```
 prometheus level=warn ts=2019-08-29T09:42:00.685168906Z caller=main.go:295 deprecation_notice="\"storage.tsdb.retention\" flag is deprecated use \"storage.tsdb.retention.time\" instead."                       │
│ prometheus level=info ts=2019-08-29T09:42:00.68527615Z caller=main.go:302 msg="Starting Prometheus" version="(version=2.7.1, branch=HEAD, revision=62e591f928ddf6b3468308b7ac1de1c63aa7fcf3)"                    │
│ prometheus level=info ts=2019-08-29T09:42:00.685304323Z caller=main.go:303 build_context="(go=go1.11.5, user=root@f9f82868fc43, date=20190131-11:16:59)"                                                         │
│ prometheus level=info ts=2019-08-29T09:42:00.685331767Z caller=main.go:304 host_details="(Linux 3.10.0-957.27.2.el7.x86_64 #1 SMP Mon Jul 29 17:46:05 UTC 2019 x86_64 prometheus-cluster-monitoring-0 (none))"   │
│ prometheus level=info ts=2019-08-29T09:42:00.68535965Z caller=main.go:305 fd_limits="(soft=1048576, hard=1048576)"                                                                                               │
│ prometheus level=info ts=2019-08-29T09:42:00.685383996Z caller=main.go:306 vm_limits="(soft=unlimited, hard=unlimited)"                                                                                          │
│ prometheus level=info ts=2019-08-29T09:42:00.68622938Z caller=main.go:620 msg="Starting TSDB ..."                                                                                                                │
│ prometheus level=info ts=2019-08-29T09:42:00.686280478Z caller=web.go:416 component=web msg="Start listening for connections" address=127.0.0.1:9090                                                             │
│ prometheus level=info ts=2019-08-29T09:42:00.686568373Z caller=main.go:489 msg="Stopping scrape discovery manager..."                                                                                            │
│ prometheus level=info ts=2019-08-29T09:42:00.686585633Z caller=main.go:503 msg="Stopping notify discovery manager..."                                                                                            │
│ prometheus level=info ts=2019-08-29T09:42:00.686596003Z caller=main.go:525 msg="Stopping scrape manager..."                                                                                                      │
│ prometheus level=info ts=2019-08-29T09:42:00.686604355Z caller=main.go:499 msg="Notify discovery manager stopped"                                                                                                │
│ prometheus level=info ts=2019-08-29T09:42:00.686630871Z caller=main.go:485 msg="Scrape discovery manager stopped"                                                                                                │
│ prometheus level=info ts=2019-08-29T09:42:00.686647036Z caller=manager.go:736 component="rule manager" msg="Stopping rule manager..."                                                                            │
│ prometheus level=info ts=2019-08-29T09:42:00.686664337Z caller=manager.go:742 component="rule manager" msg="Rule manager stopped"                                                                                │
│ prometheus level=info ts=2019-08-29T09:42:00.686680133Z caller=notifier.go:521 component=notifier msg="Stopping notification manager..."                                                                         │
│ prometheus level=info ts=2019-08-29T09:42:00.686693057Z caller=main.go:679 msg="Notifier manager stopped"                                                                                                        │
│ prometheus level=info ts=2019-08-29T09:42:00.686653567Z caller=main.go:519 msg="Scrape manager stopped"                                                                                                          │
│ prometheus level=error ts=2019-08-29T09:42:00.687009615Z caller=main.go:688 err="opening storage failed: create dir: mkdir /prometheus/wal: permission denied"
 ```

**Comment by jhu-arod** [comment]
I'm seeing the same issue with monitoring chart 0.0.3 as well.  I tried manually adding write permissions to group and other but the end result is still the same.

```
level=warn ts=2019-09-10T17:36:33.962589882Z caller=main.go:295 deprecation_notice="\"storage.tsdb.retention\" flag is deprecated use \"storage.tsdb.retention.time\" instead."
level=info ts=2019-09-10T17:36:33.96265445Z caller=main.go:302 msg="Starting Prometheus" version="(version=2.7.1, branch=HEAD, revision=62e591f928ddf6b3468308b7ac1de1c63aa7fcf3)"
level=info ts=2019-09-10T17:36:33.962680955Z caller=main.go:303 build_context="(go=go1.11.5, user=root@f9f82868fc43, date=20190131-11:16:59)"
level=info ts=2019-09-10T17:36:33.962710906Z caller=main.go:304 host_details="(Linux 4.14.138-114.102.amzn2.x86_64 #1 SMP Thu Aug 15 15:29:58 UTC 2019 x86_64 prometheus-cluster-monitoring-0 (none))"
level=info ts=2019-09-10T17:36:33.962738966Z caller=main.go:305 fd_limits="(soft=1024, hard=4096)"
level=info ts=2019-09-10T17:36:33.962761888Z caller=main.go:306 vm_limits="(soft=unlimited, hard=unlimited)"
level=info ts=2019-09-10T17:36:33.965308413Z caller=main.go:620 msg="Starting TSDB ..."
level=info ts=2019-09-10T17:36:33.965471525Z caller=web.go:416 component=web msg="Start listening for connections" address=127.0.0.1:9090
level=info ts=2019-09-10T17:36:33.965560432Z caller=main.go:489 msg="Stopping scrape discovery manager..."
level=info ts=2019-09-10T17:36:33.965572576Z caller=main.go:503 msg="Stopping notify discovery manager..."
level=info ts=2019-09-10T17:36:33.965581608Z caller=main.go:525 msg="Stopping scrape manager..."
level=info ts=2019-09-10T17:36:33.965591348Z caller=main.go:499 msg="Notify discovery manager stopped"
level=info ts=2019-09-10T17:36:33.965613126Z caller=main.go:519 msg="Scrape manager stopped"
level=info ts=2019-09-10T17:36:33.965628668Z caller=main.go:485 msg="Scrape discovery manager stopped"
level=info ts=2019-09-10T17:36:33.965644908Z caller=manager.go:736 component="rule manager" msg="Stopping rule manager..."
level=info ts=2019-09-10T17:36:33.965654212Z caller=manager.go:742 component="rule manager" msg="Rule manager stopped"
level=info ts=2019-09-10T17:36:33.965664281Z caller=notifier.go:521 component=notifier msg="Stopping notification manager..."
level=info ts=2019-09-10T17:36:33.965673714Z caller=main.go:679 msg="Notifier manager stopped"
level=error ts=2019-09-10T17:36:33.966146198Z caller=main.go:688 err="opening storage failed: create dir: mkdir /prometheus/wal: permission denied"
```

**Comment by adeleglise** [comment]
@loganhz Do you think someone could have a look on this one ?

**Comment by thxCode** [comment]
@adeleglise , @ilusharulkov , @jhu-arod , have you guys use with the `local-path` storage class? 

for this issue, you could `docker ps -a | grep prometheus` to find out the exited Prometheus container and then `docker inspect` that container ID to confirm the mounted path of `/prometheus`, for example: 
```
"Mounts": [
 ...
            {
                "Type": "bind",
                "Source": "/var/lib/kubelet/pods/3f52e135-e00c-11e9-a6de-fa163ef1e7a9/volume-subpaths/pvc-78656acf-dfb2-11e9-a6de-fa163ef1e7a9/prometheus/1",
                "Destination": "/prometheus",
                "Mode": "",
                "RW": true,
                "Propagation": "rprivate"
            },
...
]
```
make sure the source path has the permission: `1000:2000`:
```
ls /var/lib/kubelet/pods/3f52e135-e00c-11e9-a6de-fa163ef1e7a9/volume-subpaths/pvc-78656acf-dfb2-11e9-a6de-fa163ef1e7a9/prometheus/1 -al
total 12
drwxr-xr-x 3 cloud 2000 4096 Sep 26 03:23 .
drwxr-x--- 3 root  root 4096 Sep 26 03:18 ..
drwxr-xr-x 2 cloud root 4096 Sep 26 03:23 wal
```

**Comment by Leen15** [comment]
Hi, same issue here using local-path-provisioner.

**Comment by thxCode** [comment]
@adeleglise , @ilusharulkov , @jhu-arod , @Leen15, could you try to follow this workaround https://github.com/rancher/rancher/issues/14836#issuecomment-491152114 to configure your cluster? I think you don't need to mount the host root path, you just mount the target path (default is `/opt/local-path-provisioner`) of `rancher/local-path-provisioner`: https://github.com/rancher/local-path-provisioner#configuration, for example, edit your custom cluster config:
```
services:
  kubelet:
    extra_args:
      containerized: "true"
    extra_binds: 
      - "/opt/local-path-provisioner:/rootfs/opt/local-path-provisioner:rshared"
```

**Comment by granttr** [comment]
> Please let us know in a comment if you can still reproduce the issue, and we'll reopen it

As posted earlier, after this issue was closed, a similar issue has been reproduced several times. I've tested that it can be easily reproduced in monitoring versions 0.0.4 and 0.0.5 also. 

Is this problem specific to using the local-path-provisioner storageClass for persisting the monitoring data? (could be related to https://github.com/rancher/local-path-provisioner/issues/41)

Since persistence doesn't work at all when using local-path -  Should this ticket be reopened pending some fix for the local-path-provisioner? 

Thanks

**Comment by adeleglise** [comment]
I've tried with longhorn, and I can still see this bug. Using 0.0.5 version of this chart, and last version of longhorn.

**Comment by thxCode** [comment]
@adeleglise , I could not reproduce this following the below steps:

1. Setup v2.3.2 Rancher Server
1. Bring up a v1.15.5 Kubernetes cluster from 3 exiting nodes: 1 all role + 2 worker roles
1. Install v0.6.2 Longhorn App from System Project
1. Configure Prometheus instance with longhorn StorageClass during enabling cluster-level Monitoring

![image](https://user-images.githubusercontent.com/3394724/68915945-f3849100-079e-11ea-84c5-d157c27c9fd7.png)

![image](https://user-images.githubusercontent.com/3394724/68915949-f67f8180-079e-11ea-99d2-1e6041952a67.png)

**Comment by adeleglise** [comment]
@thxCode Yup I confirm. The problem was that the pvc was still here and using Local Node Path. I've deleted it, and recreate the prometheus pod. It's now working with longhorn.

**Comment by Tejeev** [comment]
We are seeing this issue with Rancher v2.3.2, Monitoring v0.0.5, and both local-path and ceph Storage Classes

**Comment by iwin4aids** [comment]
is this issue fixed in 2.4.2 release? I has met the same issue in 2.3.5

**Comment by iwin4aids** [comment]
I could alwasys reproduce this issue on rancher 2.3.5 and prometheus operator 0.0.7 .

It seems that rancher's prometheus persistent feature could not work with either k8s local storage provisioner or rancher's local node path provisioner.

Any member working on it ?

**Comment by drog** [comment]
The same issue with rancher 2.4.3 in AWS with EBS storage

logs:
```
Unable to attach or mount volumes: unmounted volumes=[
secret-exporter-etcd-cert 
configmap-prometheus-cluster-monitoring-nginx 
cluster-monitoring-token-kv6bs 
config 
nginx-home 
config-out prometheus-cluster-monitoring-db 
prometheus-cluster-monitoring-rulefiles-0
], 
unattached volumes=[
secret-exporter-etcd-cert 
configmap-prometheus-cluster-monitoring-nginx 
cluster-monitoring-token-kv6bs 
config 
nginx-home 
config-out 
prometheus-cluster-monitoring-db 
prometheus-cluster-monitoring-rulefiles-0
]: 
timed out waiting for the condition	
```

**Comment by jiaqiluo** [comment]
The bug is fixed in the latest chart `v0.1.1`
see this issue for more info: https://github.com/rancher/rancher/issues/26969
